backend/inspirehep/records/api/mixins.py

```diff
@@ -203,17 +232,106 @@ def update_refs_in_citation_table(self, save_every=100):
+            for collaboration in collaborations_values:
+                collaboration_list.append(collaboration)
+        return collaboration_list
+
+    def get_self_cited_referenced_papers(self):
+        uuid = str(self.model.id)
+        sql_query = f"""
+            SELECT papers_from_authors.record_id
+            FROM
+                (SELECT DISTINCT a1.record_id FROM {RecordsAuthors.__tablename__} a1, {RecordsAuthors.__tablename__} a2
+                WHERE
+                    a2.record_id ='{uuid}'
+                    AND a1.author_id=a2.author_id
+                    AND (
+                        (a1.id_type='INSPIRE BAI' AND a2.id_type='INSPIRE BAI')
+                         OR (a1.id_type='collaboration' AND a2.id_type='collaboration')
+                    )
+                )AS papers_from_authors,
+                (SELECT cited_id as record_id FROM
+                  {RecordCitations.__tablename__} WHERE
+                citer_id = '{uuid}'
+                UNION
+                SELECT citer_id AS record_id
+                  from {RecordCitations.__tablename__}
+                where cited_id = '{uuid}') as cited_papers
+            WHERE
+            cited_papers.record_id=papers_from_authors.record_id
+        """
+        return [str(row[0]) for row in db.session.execute(text(sql_query)).fetchall()]
+
+    def update_self_citations(self):
+        self_citations = self.get_self_cited_referenced_papers()
+        LOGGER.info(
+            "Self-citated_papers",
+            self_citations_count=len(self_citations),
+            recid=self.get("control_number"),
+        )
+        uuid = str(self.model.id)
+        if self_citations:
+            # update self-citations
+            RecordCitations.query.filter(
+                and_(
+                    or_(
+                        and_(
+                            RecordCitations.cited_id == uuid,
+                            RecordCitations.citer_id.in_(self_citations),
+                        ),
+                        and_(
+                            RecordCitations.cited_id.in_(self_citations),
+                            RecordCitations.citer_id == uuid,
+                        ),
+                    ),
+                    RecordCitations.citation_type == RecordCitationType.citation,
+                )
+            ).update(
+                {RecordCitations.citation_type: RecordCitationType.self_citation},
+                synchronize_session=False,
+            )
+        # update not-self_citations
+        RecordCitations.query.filter(
+            and_(
+                RecordCitations.cited_id == uuid,
+                not_(RecordCitations.citer_id.in_(self_citations)),
+                RecordCitations.citation_type == RecordCitationType.self_citation,
+            )
+        ).update(
+            {RecordCitations.citation_type: RecordCitationType.citation},
+            synchronize_session=False,
         )
```

Comment:
I think it would be easier to do the whole update in DB by joining on the query to get self-citations, as this is inefficient and you run into the risk of exceeding the query limit length if there are lots of self-citations.

Reply:
Making this update only in db it will make code less readable. Now it's really fast and reliable.
Still isn't the limit of PostgreSQL query a 1GB or even more? To reach near this limit we should have a paper with at least 16M self-citations.

Reply:
then why do we have the batching in https://github.com/inspirehep/inspirehep/blob/e6004615e0171eefd05778c3e222ab73161b5241/backend/inspirehep/records/api/base.py#L88-L97 ?

Reply:
Discussed IRL